src/main/java/ca/uhn/fhir/jpa/starter/CustomAuthorizationInterceptor.java

```diff
@@ -13,35 +23,107 @@
 
 @Interceptor
 public class CustomAuthorizationInterceptor extends AuthorizationInterceptor {
-
-	private static final String HEADER_NAME = "x-api-key";
-	private static final String VALIDATE_API_KEY = "VALIDATE_API_KEY";
-	private static final String API_KEY = "API_KEY";
+	private static final Logger logger = LoggerFactory.getLogger(CustomAuthorizationInterceptor.class);
+	private static final String OAUTH_URL = System.getenv("OAUTH_URL");
+	private static final String APIKEY_ENABLED = System.getenv("APIKEY_ENABLED");
+	private static final String APIKEY_HEADER = "x-api-key";
+	private static final String APIKEY = System.getenv("APIKEY");
+	private static final String FHIR_VERSION = System.getenv("fhir_version");
+	private static final String TOKEN_PREFIX = "BEARER ";
+	private static PublicKey publicKey = null;
+	private static OAuth2Helper oAuth2Helper = new OAuth2Helper();
 
 	@Override
-	public List<IAuthRule> buildRuleList(RequestDetails theRequestDetails) {
+	public List<IAuthRule> buildRuleList(RequestDetails theRequest) {
 		try {
 
-			boolean validateApiKey = Boolean.parseBoolean(System.getenv(VALIDATE_API_KEY));
-			String token = theRequestDetails.getHeader(HEADER_NAME);
+			if (theRequest.getRequestPath().equals(RestOperationTypeEnum.METADATA.getCode())) {
+				return allowAll();
+			}
+
+			if (!oAuth2Helper.isOAuthEnabled() && !isApiKeyEnabled()) {
+				logger.info("Apikey & Oauth2 are disabled");
+				return allowAll();
+			}
+
+			if (oAuth2Helper.isOAuthEnabled() && oAuth2Helper.isOAuthHeaderPresent(theRequest)) {
+				logger.info("Auhorizing via OAuth");
+				return authorizeOAuth(theRequest);
+			}
 
-			if (!validateApiKey) {
-				return authorizeRequest();
-			} else if (!StringUtils.isEmpty(token) && token.equals(System.getenv(API_KEY))) {
-				return authorizeRequest();
-			} else {				
-				return (theRequestDetails.getOperation().equals(RestOperationTypeEnum.METADATA.getCode()))  ? authorizeRequest() : denyRequest();
+			if (isApiKeyEnabled() && isApiKeyHeaderPresent(theRequest)) {
+				logger.info("Auhorizing via X-API-KEY");
+				return authorizeApiKey(theRequest);
 			}
 		} catch (Exception e) {
-			return denyRequest();
+			logger.info("Unexpected authorization error", e);
+			return denyAll();
 		}
+
+		logger.info("Authorization failure - fall through");
```

Comment:
This should probably be a warning too. It should only occur if OAuth2 and/or APIKEY is enabled and neither is specified in the request.